### ISSUE TITLE
segfault when exiting terminal

### ISSUE THREAD

**Issue by antoinemadec** [vim/vim#2899]
system:

- VIM - Vi IMproved 8.0 (2016 Sep 12, compiled May 11 2018 10:30:21)
- Included patches: 1-1813
- OS: up-to-date Ubuntu 16.04 LTS 

issue:
Vim sometimes segfault when quitting a ```terminal```. This segfault does not happen all the time and seems to be related to ```lazyredraw``` somehow.

minimal vimrc:
```vim
set nocompatible
set lazyredraw
```

steps to reproduce:

- ```:terminal```
- press ```<C-d>``` to quit as fast as you can
- this issue is random, it might take you 2-3 times before you can observe the segfault

You can find the core dump here: [core.txt](https://github.com/vim/vim/files/1996396/core.txt)
The previous version of Vim I was using was from a month ago and it was working fine.
Do you have any idea where this error is coming from ?

**Comment by dpelle** [comment]
Attached core file is useless as it can only be analyzed with your compiled vim ELF.

I cannot reproduce the issue. Can you try to reproduce it when running vim with valgrind?
```
$ valgrind --track-origins=yes --log-file=vg.log ./vim
```
vg.log file may then contain something useful if you can reproduce the bug.
It would also be nice if you can bissect to find which patch introduces the bug.

**Comment by brammool** [comment]
> system:
> 
> - VIM - Vi IMproved 8.0 (2016 Sep 12, compiled May 11 2018 10:30:21)
> - Included patches: 1-1813
> - OS: up-to-date Ubuntu 16.04 LTS 
> 
> issue:
> Vim sometimes segfault when quitting a ```terminal```. This segfault does not happen all the time and seems to be related to ```lazyredraw``` somehow.
> 
> minimal vimrc:
> ```vim
> set nocompatible
> set lazyredraw
> ```
> 
> steps to reproduce:
> 
> - ```:terminal```
> - press ```<C-d>``` to quit as fast as you can
> - this issue is random, it might take you 2-3 times before you can observe the segfault
> 
> You can find the core dump here: [core.txt](https://github.com/vim/vim/files/1996396/core.txt)
> The previous version of Vim I was using was from a month ago and it was working fine.
> Do you have any idea where this error is coming from ?

Thanks for reporting.  I managed to reproduce it under valgrind.  Looks
like the terminal window is closed while updating the screen.  Thus we
need to check if the terminal is still valid after calling
update_screen().

I'll make a patch, let me know if you can still reproduce the problem
after that.

-- 
TALL KNIGHT: When you have found the shrubbery, then you must cut down the
             mightiest tree in the forest ... with a herring.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by antoinemadec** [comment]
@dpelle , here is [vg.log](https://github.com/vim/vim/files/1996473/vg.log)
@brammool , OK, I will try again with your patch and will let you know.

Thanks a lot for the quick response!

**Comment by dpelle** [comment]
I could reproduce the bug after all, using vim-8.1814 and valgrind:
```
$ valgrind --num-callers=50 --track-origins=yes --log-file=vg.log \
  ./vim --clean -c 'set nocompatible' -c 'set lazyredraw' -c 'terminal'

==19170== Memcheck, a memory error detector
==19170== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==19170== Using Valgrind-3.14.0.GIT and LibVEX; rerun with -h for copyright info
==19170== Command: ./vim --clean -c set\ nocompatible -c set\ lazyredraw -c terminal
==19170== Parent PID: 14403
==19170== 
==19170== Invalid read of size 8
==19170==    at 0x54B429: win_redr_status.part.21 (screen.c:6895)
==19170==    by 0x5548F5: win_redr_status (screen.c:6873)
==19170==    by 0x5548F5: update_screen (screen.c:764)
==19170==    by 0x59D7B6: terminal_loop (terminal.c:2107)
==19170==    by 0x5EBAF9: main_loop (main.c:1347)
==19170==    by 0x5ECC0B: vim_main2 (main.c:915)
==19170==    by 0x410E11: main (main.c:443)
==19170==  Address 0xe3480f0 is 640 bytes inside a block of size 6,488 free'd
==19170==    at 0x4C2ECF0: free (vg_replace_malloc.c:530)
==19170==    by 0x5B7785: win_free (window.c:4682)
==19170==    by 0x5B918C: win_free_mem (window.c:2548)
==19170==    by 0x5B918C: win_close (window.c:2393)
==19170==    by 0x41C007: do_buffer (buffer.c:1428)
==19170==    by 0x41C711: do_bufdel (buffer.c:1161)
==19170==    by 0x59DE5F: term_channel_closed (terminal.c:2822)
==19170==    by 0x5E13DE: channel_close (channel.c:2947)
==19170==    by 0x5E2273: channel_close_now (channel.c:3281)
==19170==    by 0x5E2273: channel_parse_messages (channel.c:4240)
==19170==    by 0x4DA806: parse_queued_messages (misc2.c:6385)
==19170==    by 0x510BBA: mch_inchar (os_unix.c:419)
==19170==    by 0x5A2C44: ui_inchar (ui.c:190)
==19170==    by 0x4A501E: inchar (getchar.c:3070)
==19170==    by 0x4A6DC1: vgetorpeek.part.9 (getchar.c:2852)
==19170==    by 0x4A81AC: vpeekc (getchar.c:1911)
==19170==    by 0x4A81AC: char_avail (getchar.c:1908)
==19170==    by 0x54B2A0: redrawing (screen.c:10773)
==19170==    by 0x54B35F: win_redr_status.part.21 (screen.c:6883)
==19170==    by 0x5548F5: win_redr_status (screen.c:6873)
==19170==    by 0x5548F5: update_screen (screen.c:764)
==19170==    by 0x59D7B6: terminal_loop (terminal.c:2107)
==19170==    by 0x5EBAF9: main_loop (main.c:1347)
==19170==    by 0x5ECC0B: vim_main2 (main.c:915)
==19170==    by 0x410E11: main (main.c:443)
==19170==  Block was alloc'd at
==19170==    at 0x4C2DBF6: malloc (vg_replace_malloc.c:299)
==19170==    by 0x4D4920: lalloc (misc2.c:976)
==19170==    by 0x4D5254: alloc_clear (misc2.c:898)
==19170==    by 0x5B0EF3: win_alloc (window.c:4493)
==19170==    by 0x5B8897: win_split_ins (window.c:1014)
==19170==    by 0x47B48A: ex_splitview (ex_docmd.c:8271)
==19170==    by 0x59A6D8: term_start (terminal.c:447)
==19170==    by 0x59B2C7: ex_terminal (terminal.c:742)
==19170==    by 0x476AB7: do_one_cmd (ex_docmd.c:2886)
==19170==    by 0x476AB7: do_cmdline (ex_docmd.c:1040)
==19170==    by 0x5EC870: exe_commands (main.c:2937)
==19170==    by 0x5EC870: vim_main2 (main.c:812)
==19170==    by 0x410E11: main (main.c:443)
...more errors after that...
```
I can bissect later if needed.

**Comment by antoinemadec** [comment]
@brammool patch 1-1814 ( a10ae5e323f4ebe6620869f4f9b0812b78090915 ) fixed it.
Thanks a lot!

Please note that a ```make clean``` was needed to take this change into account.

**Comment by antoinemadec** [comment]
@dpelle I concur (read your message after sending mine).
Reopen issue.

**Comment by dpelle** [comment]
@antoinemadec wrote 

> patch 1-1814 ( a10ae5e ) fixed it.

Well it still crashes for me at vim-8.0.1814. I did make clean and rebuilt it.
I used:
```
$ valgrind --num-callers=50 --track-origins=yes --log-file=vg.log \
  ./vim --clean -c 'set nocp lazyredraw' -c 'terminal'
```
Then pressed CTRL-D which trigger the segfault.
I double-checked that I'm using vim-8.0.1814.

**Comment by brammool** [comment]
Domnique wrote:

> @antoinemadec wrote 
> 
> > patch 1-1814 ( a10ae5e ) fixed it.
> 
> Well it still crashes for me at vim-8.0.1814. I did make clean and rebuilt it.
> I used:
> ```
> $ valgrind --num-callers=50 --track-origins=yes --log-file=vg.log \
>   ./vim --clean -c 'set nocp lazyredraw' -c 'terminal'
> ```
> Then pressed CTRL-D which trigger the segfault.
> I double-checked that I'm using vim-8.0.1814.

Right, the patch only fixed part of the problem.  I managed to reproduce
by adding: -c 'call\ feedkeys("exit\\r")'

This needs some more work...


-- 
TIM: To the north there lies a cave,  the cave of Caerbannog, wherein, carved
     in mystic runes, upon the very living rock, the last words of Olfin
     Bedwere of Rheged make plain the last resting place of the most Holy
     Grail.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by dpelle** [comment]
This issue was fixed in patch 8.0.1815 but I don't have permission to close it.
